tests/lib/peers/PeerManager.test.ts

```diff
@@ -4,87 +4,333 @@ import Polykey from "../../../src/lib/Polykey"
+    let peerC: Polykey
+
+    beforeAll(async () => {
+      const nodes = await setupTwoConnectedNodes()
+      peerA = nodes.nodeA.pk
+      tempDirPeerA = nodes.nodeA.tempDir
+      peerB = nodes.nodeB.pk
+      tempDirPeerB = nodes.nodeB.tempDir
+
+      // open relay connection from peerB to peerA
+      await peerB.peerManager.turnClient.requestRelayConnection(peerA.peerManager.peerInfo.publicKey)
+      peerA.peerManager.updatePeer(peerB.peerManager.peerInfo)
+      peerB.peerManager.updatePeer(peerA.peerManager.peerInfo)
+      // ======== PEER C ======== //
+      // Define temp directory
+      tempDirPeerC = fs.mkdtempSync(`${os.tmpdir}/pktest${randomString()}`)
+
+      // Create keyManager
+      const keyManagerC = new KeyManager(tempDirPeerC, fs)
+      await keyManagerC.generateKeyPair('John Smith', 'john.smith@email.com', 'some passphrase', 1024, true)
+
+      // Initialize polykey
+      peerC = new Polykey(
+        tempDirPeerC,
+        fs,
+        keyManagerC
+      )
+
+      peerA.peerManager.addPeer(peerC.peerManager.peerInfo)
+      peerB.peerManager.addPeer(peerC.peerManager.peerInfo)
+
+      peerC.peerManager.addPeer(peerA.peerManager.peerInfo)
+      // remove connectedAddress from peerC to simulate a blocked connection
+      peerB.peerManager.peerInfo.connectedAddress = undefined
+      peerC.peerManager.addPeer(peerB.peerManager.peerInfo)
+    })
+
+    afterAll(() => {
+      fs.rmdirSync(tempDirPeerA, { recursive: true })
+      fs.rmdirSync(tempDirPeerB, { recursive: true })
+      fs.rmdirSync(tempDirPeerC, { recursive: true })
+    })
+
+    describe('Peer Relay Sharing', () => {
+      test('can clone a vault through a peer relay connection', async done => {
+        // ==== Pull Vault B to C ==== //
+        const vaultName = `Vault-${randomString()}`
+        const vault = await peerB.vaultManager.createVault(vaultName)
+
+        const clonedVault = await peerC.vaultManager.cloneVault(vault.name, peerB.peerManager.peerInfo.publicKey)
+        expect(vault.name).toEqual(clonedVault.name)
+
+        done()
+      })
+
+      test('can clone many vaults through a peer relay connection', async done => {
+        // ==== Pull Vaults B to C ==== //
+        const vaultNameList = [...Array(10).keys()].map((_) => {
+          return `Vault-${randomString()}`
+        })
+
+        for (const vaultName of vaultNameList) {
+          await peerB.vaultManager.createVault(vaultName)
+        }
+
+        // clone all vaults from B to C asynchronously
+        const clonedVaults = await Promise.all(vaultNameList.map(async (v) => {
+          return peerC.vaultManager.cloneVault(v, peerB.peerManager.peerInfo.publicKey)
+        }))
+        const clonedVaultNameList = clonedVaults.map((v) => {
+          return v.name
+        })
+
+        expect(clonedVaultNameList).toEqual(vaultNameList)
+
+        done()
+      }, 20000)
+    })
+  })
+
+  describe('NAT Traversal via UDP Hole Punching', () => {
```

Comment:
How are you testing this as a unit test? This is not possible without a full simulation of NAT and IP router. If this is meant to be under the NAT simulation. You may need to separate these tests entirely into a different directory under `./tests`. We will need to figure out something with NixOS and Tetra soon. One way would be for Tetra to be a program to run bigger network simulations.

Reply:
This is less a test of NAT traversal and more a test of how the code would respond if we couldn't connect directly (I mock this by setting the peerAddress to undefined)